models/labels/nft_blue_chip_holder.sql

```diff
@@ -0,0 +1,72 @@
+with blue_chip as (
+  select *
+  from {{ ref('blue_chip_collection') }}
+),
+
+erc721_transfer as (
+  select
+    contract_address as nft_contract_address,
+    tokenid as nft_token_id,
+    to as to_address,
+    evt_block_time as block_time
+  from {{ ref('ERC721_evt_Transfer') }}
+),
+
+erc1155_single_transfer as (
+  select
+    contract_address as nft_contract_address,
+    id as nft_token_id,
+    to as to_address,
+    evt_block_time as block_time
+  from {{ ref('ERC1155_evt_TransferSingle') }}
+),
+
+erc1155_batch_transfer as (
+  select
+    contract_address as nft_contract_address,
+    explode(ids) as nft_token_id,
+    to as to_address,
+    evt_block_time as block_time
+  from {{ ref('ERC1155_evt_TransferBatch') }}
+),
+
+cryptopunks_transfer as (
+  select
+    contract_address as nft_contract_address,
+    punkIndex as nft_token_id,
+    to as to_address,
+    evt_block_time as block_time
+  from {{ ref('cryptopunks_CryptoPunksMarket_evt_PunkTransfer') }}
+), 
```

Comment:
Cryptopunks market contract is very very special. According to OpenSea data, there are 3500~ holders in total. We can verify the number against that.

After some research, I have concluded the following events which will lead to the ownership of Cryptopunks.


```sql
with evt as (

    select to, punkIndex, evt_block_time
    from ethereum_cryptopunks.cryptopunksmarket_evt_assign
    union

    select toAddress as to, punkIndex, evt_block_time
    from ethereum_cryptopunks.cryptopunksmarket_evt_punkbought
    
    union 

    select to, punkIndex, evt_block_time
    from ethereum_cryptopunks.cryptopunksmarket_evt_punktransfer
)

select
    distinct to
from
    (select  
    punkIndex,
    to,
    row_number()over(partition by punkIndex order by evt_block_time desc) as rank
    from evt
     where to <> '0x0000000000000000000000000000000000000000'
    )
where rank=1
```

Reply:
`cryptopunksmarket_evt_assign` seems to be some hard-coded mint events